mono/metadata/icall.c

```diff
@@ -8523,111 +8524,105 @@ mono_lookup_icall_symbol (MonoMethod *m)
+#define ICALL_SIG_TYPE_int16 	(MonoType*)&mono_defaults.int16_class
+#define ICALL_SIG_TYPE_uint16 	(MonoType*)&mono_defaults.uint16_class
+#define ICALL_SIG_TYPE_long 	(MonoType*)&mono_defaults.int64_class
+#define ICALL_SIG_TYPE_ulong 	(MonoType*)&mono_defaults.uint64_class
+#define ICALL_SIG_TYPE_float 	(MonoType*)&mono_defaults.single_class
+#define ICALL_SIG_TYPE_double 	(MonoType*)&mono_defaults.double_class
+#define ICALL_SIG_TYPE_object 	(MonoType*)&mono_defaults.object_class
+#define ICALL_SIG_TYPE_obj 	(MonoType*)&mono_defaults.object_class
+#define ICALL_SIG_TYPE_string 	(MonoType*)&mono_defaults.string_class
+#define ICALL_SIG_TYPE_bool 	(MonoType*)&mono_defaults.boolean_class
+#define ICALL_SIG_TYPE_boolean 	(MonoType*)&mono_defaults.boolean_class
+
+// The initializer has a different type than the result and
+// is stored in the same place. Return type is with parameters.
+#define ICALL_SIG_TYPES_1(a) 		  	ICALL_SIG_TYPE_ ## a,
+#define ICALL_SIG_TYPES_2(a, b) 	  	ICALL_SIG_TYPES_1 (a            ) ICALL_SIG_TYPES_1 (b)
+#define ICALL_SIG_TYPES_3(a, b, c) 	  	ICALL_SIG_TYPES_2 (a, b         ) ICALL_SIG_TYPES_1 (c)
+#define ICALL_SIG_TYPES_4(a, b, c, d) 	  	ICALL_SIG_TYPES_3 (a, b, c      ) ICALL_SIG_TYPES_1 (d)
+#define ICALL_SIG_TYPES_5(a, b, c, d, e)	ICALL_SIG_TYPES_4 (a, b, c, d   ) ICALL_SIG_TYPES_1 (e)
+#define ICALL_SIG_TYPES_6(a, b, c, d, e, f)	ICALL_SIG_TYPES_5 (a, b, c, d, e) ICALL_SIG_TYPES_1 (f)
+#define ICALL_SIG_TYPES_7(a, b, c, d, e, f, g)	ICALL_SIG_TYPES_6 (a, b, c, d, e, f) ICALL_SIG_TYPES_1 (g)
+#define ICALL_SIG_TYPES_8(a, b, c, d, e, f, g, h) ICALL_SIG_TYPES_7 (a, b, c, d, e, f, g) ICALL_SIG_TYPES_1 (h)
+
+#define ICALL_SIG_TYPES(n, types) ICALL_SIG_TYPES_ ## n types
+
+static struct {
+#define ICALL_SIG(n, xtypes) 			\
+	struct {				\
+		MonoMethodSignature sig;	\
+		gpointer     types [n];		\
+	} ICALL_SIG_NAME (n, xtypes);
+ICALL_SIGS
+	MonoMethodSignature end;
+} mono_icall_signatures = {
+#undef ICALL_SIG
+#define ICALL_SIG(n, types) { { \
+	0,			/* ret */ \
+	n,			/* param_count */ \
+	-1,			/* sentinelpos */ \
+	0,			/* generic_param_count */ \
+	MONO_ICALL_SIGNATURE_CALL_CONVENTION, \
+	0,			/* hasthis */ \
+	0, 			/* explicit_this */ \
+	1, 			/* pinvoke */ \
+	0, 			/* is_inflated */ \
+	0,			/* has_type_parameters */ \
+}, { ICALL_SIG_TYPES (n, types) } }, /* params and ret */
+ICALL_SIGS
+};
 
-	g_strfreev (parts);
+#undef ICALL_SIG
+#define ICALL_SIG(n, types) MonoMethodSignature * const ICALL_SIG_NAME (n, types) = &mono_icall_signatures.ICALL_SIG_NAME (n, types).sig;
+ICALL_SIGS
+#undef ICALL_SIG
 
-	mono_image_lock (corlib);
-	res2 = (MonoMethodSignature *)g_hash_table_lookup (corlib->helper_signatures, sigstr);
-	if (res2)
-		res = res2; /*Value is allocated in the image pool*/
-	else
-		g_hash_table_insert (corlib->helper_signatures, (gpointer)sigstr, res);
-	mono_image_unlock (corlib);
+void
+mono_create_icall_signatures (void)
+{
```

Comment:
A lock around this function might be a good idea.
But this, and the prior code, run very early, and there are no later mutations in the prior or current code -- the array_va stuff avoids this path for example, it has its own separate synchronized cache of its own signatures (which can go away if I get the other stuff working).
I think mono_init should have a giant lock.